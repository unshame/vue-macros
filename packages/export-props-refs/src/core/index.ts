import {
  DEFINE_PROPS,
  MagicString,
  WITH_DEFAULTS,
  getTransformResult,
  isCallOf,
  parseSFC,
  importHelperFn,
} from '@vue-macros/common'
import {
  CallExpression,
  ExportNamedDeclaration,
  Expression,
  VariableDeclaration,
  type VariableDeclarator,
} from '@babel/types'

export function transformExportPropsRefs(code: string, id: string) {
  const { scriptSetup, getSetupAst } = parseSFC(code, id)
  if (!scriptSetup) return

  const offset = scriptSetup.loc.start.offset
  const s = new MagicString(code)

  const props: Record<string, { type: string; defaultValue?: string }> = {}
  let hasDefineProps = false

  function isSupported(
    declInit: Expression | null,
    declParent: VariableDeclaration
  ): declInit is CallExpression {
    return (
      declInit?.type === 'CallExpression' &&
      declInit.callee?.type === 'Identifier' &&
      declInit.callee?.name === 'ref' &&
      declParent.kind === 'const'
    )
  }

  function walkVariableDeclarator(
    decl: VariableDeclarator,
    declParent: VariableDeclaration
  ) {
    if (decl.id.type !== 'Identifier') {
      throw new Error('Only support identifier in export props refs')
    } else if (
      decl.init &&
      (isCallOf(decl.init, DEFINE_PROPS) || isCallOf(decl.init, WITH_DEFAULTS))
    ) {
      hasDefineProps = true
      return
    }

    const init = decl.init as Expression | null

    if (!isSupported(init, declParent)) {
      throw new Error(
        'Only support assigning ref<type>(defaultValue) to props in export props refs'
      )
    }

    const argument = init.arguments[0]
    const typeParam = init.typeParameters?.params[0]

    const type = typeParam ? `: ${s.sliceNode(typeParam, { offset })}` : ': any'
    const name = decl.id.name
    const defaultValue = argument
      ? s.sliceNode(argument, { offset })
      : undefined
    props[name] = { type, defaultValue }
  }

  const setupAst = getSetupAst()!
  for (const stmt of setupAst.body) {
    if (
      stmt.type === 'ExportNamedDeclaration' &&
      stmt.declaration?.type === 'VariableDeclaration'
    ) {
      for (const decl of stmt.declaration.declarations) {
        walkVariableDeclarator(decl, stmt.declaration)
      }
      s.removeNode(stmt, { offset })
    } else if (isCallOf(stmt, DEFINE_PROPS) || isCallOf(stmt, WITH_DEFAULTS)) {
      hasDefineProps = true
    }
  }

  if (Object.keys(props).length === 0) {
    return
  } else if (hasDefineProps) {
    throw new Error("Don't support export props refs mixed with defineProps")
  }

  let codegen = ''
  let destructureString = ''
  let withDefaultsString = ''
  for (const [name, { type, defaultValue }] of Object.entries(props)) {
    codegen += `  ${name}${defaultValue ? '?' : ''}${type},\n`
    destructureString += ` ${name},`
    withDefaultsString += defaultValue ? `  ${name}: ${defaultValue},\n` : ''
  }
  codegen = `withDefaults(defineProps<{\n${codegen}}>(), {\n${withDefaultsString}});
const {${destructureString} } = ${importHelperFn(
    s,
    offset,
    'toRefs'
  )}(__props);`

  s.prependLeft(offset, `\n${codegen}`)

  return getTransformResult(s, id)
}
