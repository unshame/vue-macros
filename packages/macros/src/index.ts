import { type UnpluginInstance } from 'unplugin'
import {
  type OptionsPlugin,
  type Plugin,
  type PluginType,
  createCombinePlugin,
} from 'unplugin-combine'

import { detectVueVersion } from '@vue-macros/common'
import { Devtools } from '@vue-macros/devtools'

import VueBetterDefine, {
  type Options as OptionsBetterDefine,
} from '@vue-macros/better-define'
import VueDefineEmit, {
  type Options as OptionsDefineEmit,
} from '@vue-macros/define-emit'
import VueDefineModels, {
  type Options as OptionsDefineModels,
} from '@vue-macros/define-models'
import VueDefineOptions, {
  type Options as OptionsDefineOptions,
} from 'unplugin-vue-define-options'
import VueDefineProp, {
  type Options as OptionsDefineProp,
} from '@vue-macros/define-prop'
import VueDefineProps, {
  type Options as OptionsDefineProps,
} from '@vue-macros/define-props'
import VueDefinePropsRefs, {
  type Options as OptionsDefinePropsRefs,
} from '@vue-macros/define-props-refs'
import VueDefineRender, {
  type Options as OptionsDefineRender,
} from '@vue-macros/define-render'
import VueDefineSlots, {
  type Options as OptionsDefineSlots,
} from '@vue-macros/define-slots'
import VueExportExpose, {
  type Options as OptionsExportExpose,
} from '@vue-macros/export-expose'
import VueExportProps, {
  type Options as OptionsExportProps,
} from '@vue-macros/export-props'
import VueExportPropsRefs, {
  type Options as OptionsExportPropsRefs,
} from '@vue-macros/export-props-refs'
import VueHoistStatic, {
  type Options as OptionsHoistStatic,
} from '@vue-macros/hoist-static'
import VueNamedTemplate, {
  type Options as OptionsNamedTemplate,
} from '@vue-macros/named-template'
import VueReactivityTransform, {
  type Options as OptionsReactivityTransform,
} from '@vue-macros/reactivity-transform'
import VueSetupBlock, {
  type Options as OptionsSetupBlock,
} from '@vue-macros/setup-block'
import VueSetupComponent, {
  type Options as OptionsSetupComponent,
} from '@vue-macros/setup-component'
import VueSetupSFC, {
  type Options as OptionsSetupSFC,
} from '@vue-macros/setup-sfc'
import VueShortEmits, {
  type Options as OptionsShortEmits,
} from '@vue-macros/short-emits'

export interface FeatureOptionsMap {
  betterDefine: OptionsBetterDefine
  defineEmit: OptionsDefineEmit
  defineModels: OptionsDefineModels
  defineOptions: OptionsDefineOptions
  defineProp: OptionsDefineProp
  defineProps: OptionsDefineProps
  definePropsRefs: OptionsDefinePropsRefs
  defineRender: OptionsDefineRender
  defineSlots: OptionsDefineSlots
  exportExpose: OptionsExportExpose
  exportProps: OptionsExportProps
  exportPropsRefs: OptionsExportPropsRefs
  hoistStatic: OptionsHoistStatic
  namedTemplate: OptionsNamedTemplate
  reactivityTransform: OptionsReactivityTransform
  setupBlock: OptionsSetupBlock
  setupComponent: OptionsSetupComponent
  setupSFC: OptionsSetupSFC
  shortEmits: OptionsShortEmits
}
export type FeatureName = keyof FeatureOptionsMap
export type FeatureOptions = FeatureOptionsMap[FeatureName]

export interface OptionsCommon {
  root?: string
  version?: number
  isProduction?: boolean
  plugins?: {
    vue?: any
    vueJsx?: any
  }
  /** @internal */
  nuxtContext?: {
    isClient?: boolean
  }
}

type OptionalSubOptions<T> = boolean | Omit<T, keyof OptionsCommon> | undefined

export type Options = OptionsCommon & {
  [K in FeatureName]?: OptionalSubOptions<FeatureOptionsMap[K]>
}

export type OptionsResolved = Required<OptionsCommon> & {
  [K in FeatureName]: false | FeatureOptionsMap[K]
}

export function resolveOptions({
  root,
  version,
  plugins,
  isProduction,
  nuxtContext,

  betterDefine,
  defineEmit,
  defineModels,
  defineOptions,
  defineProp,
  defineProps,
  definePropsRefs,
  defineRender,
  defineSlots,
  exportExpose,
  exportProps,
  exportPropsRefs,
  hoistStatic,
  namedTemplate,
  reactivityTransform,
  setupBlock,
  setupComponent,
  setupSFC,
  shortEmits,
}: Options): OptionsResolved {
  function resolveSubOptions<K extends FeatureName>(
    options: OptionalSubOptions<FeatureOptionsMap[K]>,
    commonOptions: Required<
      Pick<OptionsCommon, keyof OptionsCommon & keyof FeatureOptionsMap[K]>
    >,
    defaultEnabled = true
  ): FeatureOptionsMap[K] | false {
    options = options ?? defaultEnabled
    if (!options) return false
    return { ...(options === true ? {} : options), ...commonOptions }
  }

  root = root || process.cwd()
  version = version || detectVueVersion(root)
  isProduction = isProduction ?? process.env.NODE_ENV === 'production'

  return {
    plugins: plugins || {},
    root,
    version,
    isProduction,
    nuxtContext: nuxtContext || {},

    betterDefine: resolveSubOptions<'betterDefine'>(betterDefine, {
      version,
      isProduction,
    }),
    defineEmit: resolveSubOptions<'defineEmit'>(defineEmit, {
      isProduction,
      version,
    }),
    defineModels: resolveSubOptions<'defineModels'>(defineModels, { version }),
    defineOptions: resolveSubOptions<'defineOptions'>(
      defineOptions,
      { version },
      version < 3.3
    ),
    defineProp: resolveSubOptions<'defineProp'>(defineProp, {
      isProduction,
      version,
    }),
    defineProps: resolveSubOptions<'defineProps'>(defineProps, { version }),
    definePropsRefs: resolveSubOptions<'definePropsRefs'>(definePropsRefs, {
      version,
    }),
    defineRender: resolveSubOptions<'defineRender'>(defineRender, { version }),
    defineSlots: resolveSubOptions<'defineSlots'>(
      defineSlots,
      { version },
      version < 3.3
    ),
    exportExpose: resolveSubOptions<'exportExpose'>(
      exportExpose,
      { version },
      false
    ),
    exportProps: resolveSubOptions<'exportProps'>(
      exportProps,
      { version },
      false
    ),
    exportPropsRefs: resolveSubOptions<'exportPropsRefs'>(
      exportPropsRefs,
      { version },
      false
    ),
    hoistStatic: resolveSubOptions<'hoistStatic'>(hoistStatic, { version }),
    namedTemplate: resolveSubOptions<'namedTemplate'>(namedTemplate, {
      version,
    }),
    reactivityTransform: resolveSubOptions<'reactivityTransform'>(
      reactivityTransform,
      { version }
    ),
    setupBlock: resolveSubOptions<'setupBlock'>(setupBlock, { version }, false),
    setupComponent: resolveSubOptions<'setupComponent'>(setupComponent, {
      version,
      root,
    }),
    setupSFC: resolveSubOptions<'setupSFC'>(setupSFC, { version }),
    shortEmits: resolveSubOptions<'shortEmits'>(
      shortEmits,
      { version },
      version < 3.3
    ),
  }
}

function resolvePlugin(
  unplugin: UnpluginInstance<any, true>,
  framework: PluginType,
  options: FeatureOptions | false
): Plugin[] | undefined

function resolvePlugin(
  unplugin: UnpluginInstance<any, false>,
  framework: PluginType,
  options: FeatureOptions | false
): Plugin | undefined

function resolvePlugin(
  unplugin: UnpluginInstance<any, boolean>,
  framework: PluginType,
  options: FeatureOptions | false
): Plugin | Plugin[] | undefined {
  if (!options) return
  return unplugin[framework!](options)
}

const name = 'unplugin-vue-macros'

export default createCombinePlugin<Options | undefined>(
  (userOptions = {}, meta) => {
    const options = resolveOptions(userOptions)

    const framework = meta.framework!
    const setupComponentPlugins = resolvePlugin(
      VueSetupComponent,
      framework,
      options.setupComponent
    )
    const namedTemplatePlugins = resolvePlugin(
      VueNamedTemplate,
      framework,
      options.namedTemplate
    )

    const plugins: OptionsPlugin[] = [
      resolvePlugin(VueSetupSFC, framework, options.setupSFC),
      setupComponentPlugins?.[0],
      resolvePlugin(VueSetupBlock, framework, options.setupBlock),
      namedTemplatePlugins?.[0],
      resolvePlugin(VueDefineProp, framework, options.defineProp),
      resolvePlugin(VueDefineEmit, framework, options.defineEmit),
      resolvePlugin(VueDefineProps, framework, options.defineProps),
      resolvePlugin(VueDefinePropsRefs, framework, options.definePropsRefs),
      resolvePlugin(VueExportProps, framework, options.exportProps),
      resolvePlugin(VueExportPropsRefs, framework, options.exportPropsRefs),
      resolvePlugin(VueShortEmits, framework, options.shortEmits),
      resolvePlugin(VueDefineModels, framework, options.defineModels),
      resolvePlugin(VueDefineSlots, framework, options.defineSlots),
      resolvePlugin(VueExportExpose, framework, options.exportExpose),
      resolvePlugin(
        VueReactivityTransform,
        framework,
        options.reactivityTransform
      ),
      resolvePlugin(VueBetterDefine, framework, options.betterDefine),
      resolvePlugin(VueHoistStatic, framework, options.hoistStatic),
      resolvePlugin(VueDefineOptions, framework, options.defineOptions),
      options.plugins.vue,
      options.plugins.vueJsx,
      resolvePlugin(VueDefineRender, framework, options.defineRender),
      setupComponentPlugins?.[1],
      namedTemplatePlugins?.[1],
      framework === 'vite'
        ? Devtools({ nuxtContext: options.nuxtContext })
        : undefined,
    ].filter(Boolean)

    return {
      name,
      plugins,
    }
  }
)
