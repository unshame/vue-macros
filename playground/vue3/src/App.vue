<script setup lang="ts">
import DefineRenderVue from './examples/define-render/index.vue'
import DefineOptionsVue from './examples/define-options/index.vue'
import DefinePropsVue from './examples/define-props/index.vue'
import DefinePropsRefsVue from './examples/define-props-refs/index.vue'
import HoistStaticVue from './examples/hoist-static/index.vue'
import DefineModelReactivityVue from './examples/define-models/reactivity-transform/parent.vue'
import DefineModelRuntimeVue from './examples/define-models/runtime/parent.vue'
import DefineSlotsVue from './examples/define-slots/parent.vue'
import ExportPropsVue from './examples/export-props/index.vue'
import ExportPropsRefsVue from './examples/export-props-refs/index.vue'
import ExportExposeVue from './examples/export-expose/index.vue'
import BetterDefineVue from './examples/better-define/index.vue'
// import NamedTemplateVue from './examples/named-template/index.vue'
import SetupBlockVue from './examples/setup-block/index.vue'

import DefinePropVue from './examples/define-prop/index.vue'
import DefineEmitVue from './examples/define-emit/index.vue'

import { SetupComponentFoo } from './examples/setup-component'

import SetupSFC from './examples/setup-sfc/index.setup'
import ShortVmodel from './examples/short-vmodel/parent.vue'
import Full from './examples/full.setup'
</script>

<template>
  <div grid="~ cols-3 gap2">
    <fieldset>
      <legend>defineOptions</legend>
      <DefineOptionsVue />
    </fieldset>

    <fieldset>
      <legend>defineRender</legend>
      <DefineRenderVue />
    </fieldset>

    <fieldset>
      <legend>defineProps</legend>
      <DefinePropsVue />
    </fieldset>

    <fieldset>
      <legend>definePropsRefs</legend>
      <DefinePropsRefsVue />
    </fieldset>

    <fieldset>
      <legend>exportProps</legend>
      <ExportPropsVue />
    </fieldset>

    <fieldset>
      <legend>exportPropsRefs</legend>
      <ExportPropsRefsVue />
    </fieldset>

    <fieldset>
      <legend>exportExpose</legend>
      <ExportExposeVue />
    </fieldset>

    <fieldset>
      <legend>hostStatic</legend>
      <HoistStaticVue />
    </fieldset>

    <fieldset>
      <legend>namedTemplate</legend>
      <!-- <NamedTemplateVue /> -->
      TODO
    </fieldset>

    <fieldset>
      <legend>defineModels</legend>
      <DefineModelReactivityVue />
      <hr />
      <DefineModelRuntimeVue />
    </fieldset>

    <fieldset>
      <legend>defineSlots</legend>
      <DefineSlotsVue />
    </fieldset>

    <fieldset>
      <legend>betterDefine</legend>
      <BetterDefineVue />
    </fieldset>

    <fieldset>
      <legend>setupComponent</legend>
      <SetupComponentFoo />
    </fieldset>

    <fieldset>
      <legend>setupSFC</legend>
      <SetupSFC />
    </fieldset>

    <fieldset>
      <legend>setupBlock</legend>
      <SetupBlockVue />
    </fieldset>

    <fieldset>
      <legend>Short v-model</legend>
      <short-vmodel />
    </fieldset>

    <fieldset>
      <legend>Full example</legend>
      <Full model-value="123" />
    </fieldset>

    <fieldset>
      <legend>defineProp</legend>
      <DefinePropVue />
    </fieldset>

    <fieldset>
      <legend>defineEmit</legend>
      <DefineEmitVue />
    </fieldset>
  </div>
</template>

<style>
legend {
  font-family: monospace;
  font-size: 16px;
  font-weight: bold;
}
</style>
