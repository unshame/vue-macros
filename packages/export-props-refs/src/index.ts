import { createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  type MarkRequired,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { transformExportPropsRefs } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOption(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC, REGEX_VUE_SUB],
    ...options,
    version,
  }
}

const name = 'unplugin-vue-export-props-refs'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}) => {
    const options = resolveOption(userOptions)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformExportPropsRefs(code, id)
      },
    }
  }
)
