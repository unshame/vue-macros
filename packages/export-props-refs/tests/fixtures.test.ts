import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformExportPropsRefs } from '../src/core'

describe('fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/*.{vue,js,ts}', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) => transformExportPropsRefs(code, id)?.code
  )
})
