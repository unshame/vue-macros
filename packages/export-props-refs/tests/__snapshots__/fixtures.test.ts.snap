// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > ./fixtures/basic.vue 1`] = `
"<script setup lang=\\"ts\\">
withDefaults(defineProps<{
  foo?: string,
  bar?: any,
  qux?: any,
}>(), {
  foo: 'foo',
  bar: 10,
  qux: fn(),
});
const { foo, bar, qux, } = __MACROS_toRefs(__props);
import { toRefs as __MACROS_toRefs } from \\"vue\\";
import { ref } from 'vue'





function fn() {}
</script>
"
`;
